### PR TITLE
Implement Yul IR generation for abi.encode*

### PR DESCRIPTION
**Pull request by chriseth**
Part of #8343

### CODE REVIEW

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 1]
```diff
@@ -730,6 +730,85 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
...
+	{
+		bool const isPacked = functionType->kind() == FunctionType::Kind::ABIEncodePacked;
+		solAssert(functionType->padArguments() != isPacked, "");
+		bool const hasSelectorOrSignature =
+			functionType->kind() == FunctionType::Kind::ABIEncodeWithSelector ||
+			functionType->kind() == FunctionType::Kind::ABIEncodeWithSignature;
+
+		TypePointers argumentTypes;
+		TypePointers targetTypes;
+		for (unsigned i = 0; i < arguments.size(); ++i)
```
Comment (mijovic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
		for (size_t i = 0; i < arguments.size(); ++i)
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 2]
```diff
@@ -730,6 +730,85 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
...
+					"))\n";
+				IRVariable selectorVariable(m_context.newYulVariable(), *TypeProvider::fixedBytes(4));
+				define(selectorVariable, hashVariable);
+			}
+		}
+		else if (functionType->kind() == FunctionType::Kind::ABIEncodeWithSelector)
+			selector = convert(*arguments.front(), *TypeProvider::fixedBytes(4)).name();
+
+		Whiskers templ(R"(
+			let <data> := <allocateTemporary>()
```
Comment (mijovic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:845]: I think allocating temporary memory should be paired with releasing it

Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:845]: It cannot be released. This is an example where we need to allocate a piece of memory where we only knew the length later on. Actually I fully forgot to do that, i.e. update the free memory pointer.

Comment (mijovic) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:845]: Ah sorry, I didn't look carefully.
But, as you said in this case updating free pointer is needed

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 3]
```diff
@@ -730,6 +730,85 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
...
+		solAssert(functionType->padArguments() != isPacked, "");
+		bool const hasSelectorOrSignature =
+			functionType->kind() == FunctionType::Kind::ABIEncodeWithSelector ||
+			functionType->kind() == FunctionType::Kind::ABIEncodeWithSignature;
+
+		TypePointers argumentTypes;
+		TypePointers targetTypes;
+		for (size_t i = 0; i < arguments.size(); ++i)
+		{
+			// ignore selctor
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
			// ignore selector
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 4]
```diff
@@ -761,6 +840,7 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 			"(" <<
 			array.commaSeparatedList() <<
 			"))\n";
+		// TODO could reset free memory pointer here
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: In this PR or later later?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 5]
```diff
@@ -730,6 +730,85 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
...
+			// hash the signature
+			Type const& selectorType = type(*arguments.front());
+			if (auto const* stringType = dynamic_cast<StringLiteralType const*>(&selectorType))
+			{
+				FixedHash<4> hash(keccak256(stringType->value()));
+				selector = formatNumber(u256(FixedHash<4>::Arith(hash)) << (256 - 32));
+			}
+			else
+			{
+				// TODO could reset free memory pointer later
```
Comment (leonardoalt) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: In this PR or later later?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:845]: Looks like this is still needed

**Comment by ekpyron** [comment]
I find it rather suspicious that ``t_ubu_soltest_enforce_yul`` doesn't hit...

**Comment by chriseth** [comment]
@ekpyron me, too. I'll dig into it!

**Comment by mijovic** [comment]
Looks like more tests can be activated with compileViaYul

**Comment by chriseth** [comment]
Let's merge #8863 first and see what else we can activate.

**Review by ekpyron — APPROVED** [review]
It's quite reassuring that all the calldata tests kick in right away :-).

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 6]
```diff
@@ -781,6 +781,94 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
...
+		bool const isPacked = functionType->kind() == FunctionType::Kind::ABIEncodePacked;
+		solAssert(functionType->padArguments() != isPacked, "");
+		bool const hasSelectorOrSignature =
+			functionType->kind() == FunctionType::Kind::ABIEncodeWithSelector ||
+			functionType->kind() == FunctionType::Kind::ABIEncodeWithSignature;
+
+		TypePointers argumentTypes;
+		TypePointers targetTypes;
+		vector<string> argumentVars;
+		for (size_t i = 0; i < arguments.size(); ++i)
```
Comment (ekpyron) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:798]: With ranges-v3:

```
for (auto const& argument: arguments | views::drop(hasSelectorOrSignature ? 1 : 0))
```
I believe ;-).

Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:798]: No sure I find that notation very intuitive...

**Pull request merged**